rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+- [`ad41e043760e`](https://github.com/NixOS/nixpkgs/commit/ad41e043760ed1da3d8c957b3bf168bdfc9bd9e2) [@jonringer](https://github.com/jonringer) - python310Packages.moto: disable failing tests after werkzeug update
+- [`d6d2d6c6d7fc`](https://github.com/NixOS/nixpkgs/commit/d6d2d6c6d7fcd5b71e97b9ee5f25c72f30eb9127) [@mweinelt](https://github.com/mweinelt) - python3Packages.twisted: skip failing tests on aarch64-darwin
+
+> **Note**
+> This was generated with a fairly hacky and non-reusable script, but it can relatively easily be verified probabilistically by picking a random commit in the time range and checking if it belongs to a pull request, repeat to increase confidence.
+
+</details>
+
+## Emergency changes
+
+Sometimes channels have blocking breakages and need to be fixed as soon as possible (citation needed).
+Currently this can be done with direct pushes, but a pull request will be required with this proposal.
+The time required to fix such breakages however is barely affected: Since there is currently no requirement for pull requests to be approved or pass CI, they can get merged immediately after opening if necessary.
+
+## Staging workflow
+
+The staging workflow is not affected because it [already uses pull requests](https://github.com/NixOS/nixpkgs/pull/241951) for all merges into the affected branches.
+
+## Direct push detection
+
+There is a [GitHub Actions workflow to detect directly pushed commits](https://github.com/NixOS/nixpkgs/blob/0b411c1e040870e89a3e598437e708979137b665/.github/workflows/direct-push.yml).
+When detected, it creates a comment in the commit pointing out that this is discouraged and linking to [this issue](https://github.com/NixOS/nixpkgs/issues/118661), where it's easy to see all direct pushes.
+The script occasionally has false positives ([issue](https://github.com/NixOS/nixpkgs/issues/240314), which creates some unnecessary commit comment notifications.
+
+With this proposal accepted it won't be possible to directly push commits anymore, making that workflow unnecessary.
+It can be removed and the above two issues can be closed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Even trivial changes that don't need a review now require a pull request, which takes more effort and creates noise for reviewers.
+- It takes slightly longer to push a quick fix in emergency situations.
+
+# Alternatives
+[alternatives]: #alternatives
+
+# Prior art
+[prior-art]: #prior-art
+
+The previous [RFC 79](https://github.com/NixOS/rfcs/pull/79) attempted to do the same, but:
+- It had a mistaken estimate for the percentage of direct master commits, calculating it to be 46.85% in the last year.
+  The mistake was assuming that all non-merge commits were direct pushes.
+  This made it seem like the change was much more impactful than it actually would've been.
+- It was too ambitious by also proposing to require accepting reviews for all pull requests.
```

Comment:
I think another important point is that in the years since #79 the discussions lead to various workflows to migrate gradually away from direct pushes.

Reply:
It takes a bit of time, but I can also check how many commits were directly pushed to master back then, I wouldn't be surprised if it was already very low.

Reply:
It was pretty low for sure, but noticeably higher — and the updates to `staging` workflows in the meantime probably matter.

Reply:
@7c6f434c are you saying that customs changed between #79 and now? (If so I would be curious why). Or was there something else that makes the change more palatable now than it was back then?

Reply:
Back then, the majority of changes already went through a PR at some point on their way to the main branch. For the exceptions, there were a few years of advertising specific changes to the specific workflows while talking up the virtues of ofBorg eval check until everyone either agrees or is nagged into compliance; apparently after all the workflow work the _strong_ arguments for direct pushes to the channel-feeding branches are resolved.

This is the difference between «now» and «the current RFC transplanted back in time with the data re-processed according to the moment of submission». «RFC 79, try again» would hopefully fail again, because of the other demands it also had.

Reply:
I got the results, out of a total 50457 master commits in the last year at RFC 79 start time [^1], out of a total 50457 commits, at most 2682 of them (5.315%) were direct pushes to master, complete listing [here](https://gist.github.com/infinisil/9e7da387082c165789b963b935890bf2) (not counting the 153 obvious false positives from @web-flow, see https://github.com/NixOS/rfcs/pull/156#discussion_r1271288179).

So while it was still about 10 times less than the original estimate in the RFC (46.85%), it was 100 times more than now (0.0517%)! I will update the RFC with this information.

[^1]: Unix epoch 1569369600 to 1600992000